dbengine/elasticsearch_engine.go

```diff
@@ -6,61 +6,46 @@ import (
 	"regexp"
 	"strings"
 
-	"github.com/alphagov/paas-compose-broker/client/elastic"
 	composeapi "github.com/compose/gocomposeapi"
 )
 
 type ElasticSearchEngine struct {
-	credentials *Credentials
+	deployment *composeapi.Deployment
 }
 
-func NewElasticSearchEngine() *ElasticSearchEngine {
-	return &ElasticSearchEngine{}
+func NewElasticSearchEngine(deployment *composeapi.Deployment) *ElasticSearchEngine {
+	return &ElasticSearchEngine{deployment}
 }
 
-func (e *ElasticSearchEngine) CreateUser(instanceID, bindingID string, deployment *composeapi.Deployment) (*Credentials, error) {
-	return e.credentials, nil
-}
-
-func (e *ElasticSearchEngine) DropUser(instanceID, bindingID string, deployment *composeapi.Deployment) error {
-	return nil
-}
-
-func (e *ElasticSearchEngine) Open(creds *Credentials) error {
-	if creds == nil {
-		return fmt.Errorf("credentials: not provided")
-	}
-
-	e.credentials = creds
-
-	err := e.testConnection()
-	if err != nil {
-		return fmt.Errorf("connection refused: %s", err)
-	}
-
-	return nil
-}
-
-func (e *ElasticSearchEngine) Close() {}
-
-func (e *ElasticSearchEngine) ParseConnectionString(deployment *composeapi.Deployment) (*Credentials, error) {
-	if deployment == nil {
+func (e *ElasticSearchEngine) GenerateCredentials(instanceID, bindingID string) (*Credentials, error) {
+	if e.deployment == nil {
 		return nil, fmt.Errorf("no deployment provided: cannot parse the connection string")
+	} else if len(e.deployment.Connection.Direct) < 1 {
+		return nil, fmt.Errorf("failed to get connection string")
 	}
 
-	u, err := url.Parse(deployment.Connection.Direct[0])
+	u, err := url.Parse(e.deployment.Connection.Direct[0])
 	if err != nil {
 		return nil, err
+	} else if u.User == nil {
+		return nil, fmt.Errorf("connection string did not contain a user")
 	}
 
+	// FIXME: Follow up story should fix connection string handling.
```

Comment:
Was this copied from MongoDB? Did the tests expose a problem with it not being here, or was it a potential problem you noticed during refactoring? Could you describe it in a commit message (and possibly a separate commit)?

Reply:
Yep cool I'll pick it out into a separate commit.

The previous "unit" test for elasticsearch named `should successfully ParseConnectionString() with more than one host` did not have an assertion for checking the Port returned in the credentials .... we added an assertion for the port in the new tests which surfaced the issue and required a similar fix.

I think that this is _really_ an issue with the Credentials structure (which assumes a single host).... a better structure for all returned compose services might be:

```
{
    URI: "db://x:y@host1.com:8080,host2.com:8181/blah",
    Addrs: ["host1.com:8080", "host2.com:8181"],
    Name: "blah"
    Username: "x",
    Password: "y",
}
```

...but obviously this would be a breaking change to any existing bindings... might be something to think about crowbar-ing in while still in "alpha/beta"?

Reply:
On second thought... I settled for extending the commit message of the refactor commit to highlight the reason for the FIXME/workaround.

Breaking out into a second commit would just leave the previous commit with the bug and add little value.

Reply:
👌  I'm going to add a note to [#147258015](https://www.pivotaltracker.com/story/show/147258015) where we were going to fix this for MongoDB, to say that it affects Elasticsearch too.